ArtOfIllusion/src/artofillusion/ScrollViewTool.java

```diff
@@ -0,0 +1,301 @@
+/* Copyright (C) 2017 by Petri Ihalainen
```

Comment:
> A ViewerCanvas should not be forcibly drawn on if you are interacting with different ViewerCanvas. I'd almost wonder if a better option would be to temporarily add an object - an actual 3d wireframe - into the scene.

Well, actually ViewerCanvases are not drawing on eachothers. Each viewer canvas has a drawable public viewcone "object" built-in and that is then just given values/activated/deactivated by the tool, that has acces to the parent window. Then at drawOverlay each ViewerCanvas calls a method to draw the "object", if the object is existing. (Using renderLine of the CanvasDrawers.)

That seemed to make better sense, than drawing anything temporary on the scene, deciding which views to draw in on and then deleting it each time a Camera is moving as the same information is needed "all the time".

At he moment the "object" is just a group of Vec3Ds with specific names and the method that draws the view cone knows, what to make of them. That of course is a very prototype kind of contraption at the moment. I'm thinking, that the "externalGraphicObject" could be more like a list of things to draw, that is then used to display any temporary information about things that are not happening on that particilar ViewerCanvas.

To me it seemd better to put the function that uses the graphics object into a tool/handler/something that is created along with the parent window, rather than code it into every subclass of the ViewerCanvas sepatrately. At the moment all the "tools" (Scroll, Move, Rotate) contain the same piece of code that uses the graphics "object" of the ViewerCanvases of "theWindow".

> By click only, I mean that they should only have Icons. They should be stripped of the "MetaTool" and "AltTool" arrangement. They would have the functions that they currently have in the 3.0.3 master branch. (In fact, they might even be removed entirely in the future.) The improved navigation tools that you are developing here would be an entirely separate implementation.

Yes, this is very much the situation with all the 3D-software I have ever used. The move and rotate tools for the view are usually existing and found as icons in the toolbars/palettes but they are hardly ever used as the same things and much more are found under mouse buttons. Of course I have no idea how the software behind there is written. In some ways the icon activated tools seem a bit like a remnant from the times, when mouses and mouse handling in the software were not that developed but so do many other things, so let's not go removing those just yet. :) -- And AFAIK any travel modes in view handling were not existing 17 years ago. The first ones, I saw, came along a few years later and BOY, were they hard to handle!

> Would the proposed mouseHandler ever make sense outside the context of a ViewerCanvas? If not, it should remain in the ViewerCanvas.

Hmm... Isn't that, what the situation is with the MoveViewTool and RotateViewTool at the moment? They are only used with ViewerCanvas though they are just added to every sub class at the creation of the parent window. I guess that should provide the option to use the the ViewerCanvas in contexts, where the user is not supposed to move or rotate the view.

Any way, so far, of course, I have been mostly focusing on creting the features, that I think should be included. -- To use mechanical terms, it could be called a "functional prototype" and the final compositon that can be "released for manufacturing" has to be figured out after we know what the machine needs to work and what the excact supply chain is....   -- From that world, I have picked the idea, that generally speaking a relatively high level of modularity is a good thing, when it comes to keeping highly complex things under control and to some extent flexible :)